hoomd/mpcd/ReverseNonequilibriumShearFlow.cc

```diff
@@ -0,0 +1,377 @@
+
+    // stage swaps into queue
+    ArrayHandle<Scalar2> h_particles_lo(m_particles_lo, access_location::host, access_mode::read);
+    ArrayHandle<Scalar2> h_particles_hi(m_particles_hi, access_location::host, access_mode::read);
+    ArrayHandle<Scalar2> h_particles_staged(m_particles_staged,
+                                      access_location::host,
+                                      access_mode::overwrite);
+    m_num_staged = 0;
+    unsigned int lo_idx = 0;
+    unsigned int hi_idx = 0;
+    for (unsigned int i = 0; i < num_pairs; ++i)
+        {
+#ifdef ENABLE_MPI
+        if (m_sysdef->isDomainDecomposed())
+            {
+            const auto mpi_comm = m_exec_conf->getMPICommunicator();
+            const int rank = m_exec_conf->getRank();
+
+            // find particle in lo slab closest to +p on all ranks
+            Scalar_Int lo_swap;
+            lo_swap.s = (layer_lo_idx < m_num_lo)
+                            ? std::fabs(h_layer_lo.data[layer_lo_idx].y - m_target_momentum)
+                            : std::numeric_limits<Scalar>::max();
+            lo_swap.i = rank;
+            MPI_Allreduce(MPI_IN_PLACE, &lo_swap, 1, MPI_HOOMD_SCALAR_INT, MPI_MINLOC, mpi_comm);
+
+            // find particle in hi slab closest to -p on all ranks
+            Scalar_Int hi_swap;
+            hi_swap.s = (layer_hi_idx < m_num_hi)
+                            ? std::fabs(h_layer_hi.data[layer_hi_idx].y + m_target_momentum)
+                            : std::numeric_limits<Scalar>::max();
+            hi_swap.i = rank;
+            MPI_Allreduce(MPI_IN_PLACE, &hi_swap, 1, MPI_HOOMD_SCALAR_INT, MPI_MINLOC, mpi_comm);
+
+            if (lo_swap.i == rank || hi_swap.i == rank) // at most, 2 ranks participate in the swap
+                {
+                if (lo_swap.i != hi_swap.i) // particles to swap are on different ranks, needs MPI
+                    {
+                    Scalar2 particle;
+                    unsigned int dest;
+                    if (lo_swap.i == rank)
+                        {
+                        particle = h_layer_lo.data[layer_lo_idx++];
+                        dest = hi_swap.i;
+                        }
+                    else
+                        {
+                        particle = h_layer_hi.data[layer_hi_idx++];
+                        dest = lo_swap.i;
+                        }
+
+                    MPI_Sendrecv_replace(&particle.y,
+                                         1,
+                                         MPI_HOOMD_SCALAR,
+                                         dest,
+                                         0,
+                                         dest,
+                                         0,
+                                         mpi_comm,
+                                         MPI_STATUS_IGNORE);
+
+                    h_swap_queue.data[m_num_staged++] = particle;
```

Comment:
All instances of this need to be replaced. I'm guessing you didn't actually compile this? It will fail because of this.
```suggestion
                    h_particles_staged.data[m_num_staged++] = particle;
```